src/IO/S3/Client.h

```diff
@@ -0,0 +1,309 @@
+#pragma once
+
+#include "config.h"
+
+#if USE_AWS_S3
+
+#include <Common/logger_useful.h>
+#include <Common/assert_cast.h>
+#include <base/scope_guard.h>
+
+#include <IO/S3/URI.h>
+#include <IO/S3/Requests.h>
+
+#include <aws/core/client/DefaultRetryStrategy.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/S3ServiceClientModel.h>
+#include <aws/core/client/AWSErrorMarshaller.h>
+#include <aws/core/client/RetryStrategy.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int TOO_MANY_REDIRECTS;
+}
+
+namespace S3
+{
+
+namespace Model = Aws::S3::Model;
+
+struct ClientCache
+{
+    ClientCache() = default;
+
+    ClientCache(const ClientCache & other)
+        : region_for_bucket_cache(other.region_for_bucket_cache)
+        , uri_for_bucket_cache(other.uri_for_bucket_cache)
+    {}
+
+    ClientCache(ClientCache && other) = delete;
+
+    ClientCache & operator=(const ClientCache &) = delete;
+    ClientCache & operator=(ClientCache &&) = delete;
+
+    void clearCache();
+
+    std::mutex region_cache_mutex;
+    std::unordered_map<std::string, std::string> region_for_bucket_cache;
+
+    std::mutex uri_cache_mutex;
+    std::unordered_map<std::string, URI> uri_for_bucket_cache;
+};
+
+class ClientCacheRegistry
+{
+public:
+    static ClientCacheRegistry & instance()
+    {
+        static ClientCacheRegistry registry;
+        return registry;
+    }
+
+    void registerClient(const std::shared_ptr<ClientCache> & client_cache);
+    void unregisterClient(ClientCache * client);
+    void clearCacheForAll();
+private:
+    ClientCacheRegistry() = default;
+
+    std::mutex clients_mutex;
+    std::unordered_map<ClientCache *, std::weak_ptr<ClientCache>> client_caches;
+};
+
+/// Client that improves the client from the AWS SDK
+/// - inject region and URI into requests so they are rerouted to the correct destination if needed
+/// - automatically detect endpoint and regions for each bucket and cache them
+///
+/// For this client to work correctly both Client::RetryStrategy and Requests defined in <IO/S3/Requests.h> should be used.
+class Client : public Aws::S3::S3Client
+{
+public:
+    template <typename... Args>
+    static std::unique_ptr<Client> create(Args &&... args)
```

Comment:
We have only one or two places where `S3::Client` is constructed, so we can use concrete a list of specific argument names and types here. Variadic templates make debugging worse, I'd prefer not to use them until it's really necessary.